vmm/src/landlock.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright © 2024 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+use crate::{Deserialize, Serialize};
+use bitflags::bitflags;
+use thiserror::Error;
+
+bitflags! {
+    #[derive(Clone, Copy, Debug, PartialEq, Eq,  Serialize, Deserialize)]
+    pub struct Perms: u8 {
+        const READ = 0b00000001;
+        const WRITE = 0b00000010;
+        const EXECUTE = 0b00000100;
+    }
+}
+use crate::vm_config::LandlockConfig;
+use landlock::{
+    path_beneath_rules, Access, AccessFs, BitFlags, Ruleset, RulesetAttr, RulesetCreated,
+    RulesetCreatedAttr, RulesetError, ABI,
+};
+use std::io::Error as IoError;
+use std::path::PathBuf;
+
+#[derive(Debug, Error)]
+pub enum LandlockError {
+    /// All RulesetErrors from Landlock library are wrapped in this error
+    #[error("Error creating/adding/restricting ruleset: {0}")]
+    ManageRuleset(#[source] RulesetError),
+
+    /// Error opening path
+    #[error("Error opening path: {0}")]
+    OpenPath(#[source] IoError),
+
+    /// Invalid Path
+    #[error("Invalid path")]
+    InvalidPath,
+}
+// https://docs.rs/landlock/latest/landlock/enum.ABI.html for more info on ABI
+static ABI: ABI = ABI::V3;
+
+pub struct Landlock {
+    ruleset: RulesetCreated,
+}
+
+impl Landlock {
+    pub fn new() -> Result<Landlock, LandlockError> {
+        let file_access = AccessFs::from_all(ABI);
+
+        let def_ruleset = Ruleset::default()
+            .handle_access(file_access)
+            .map_err(LandlockError::ManageRuleset)?;
+
+        // By default, ruelsets are created in `BestEffort` mode. This lets Landlock
+        // to enable all the supported rules and silently ignore the unsupported ones.
+        let ruleset = def_ruleset.create().map_err(LandlockError::ManageRuleset)?;
+
+        Ok(Landlock { ruleset })
+    }
+
+    pub fn add_rule(
+        &mut self,
+        path: PathBuf,
+        access: BitFlags<AccessFs>,
+    ) -> Result<(), LandlockError> {
+        // path_beneath_rules in landlock crate handles file and directory access rules.
+        // Incoming path/s are passed to path_beneath_rules, so that we don't
+        // have to worry about the type of the path.
+        let paths = vec![path.clone()];
+        let path_beneath_rules = path_beneath_rules(paths, access);
+        self.ruleset
+            .as_mut()
+            .add_rules(path_beneath_rules)
+            .map_err(LandlockError::ManageRuleset)?;
+        Ok(())
+    }
+
+    fn perms_to_access(&self, perms: Perms) -> BitFlags<AccessFs> {
+        let mut access = BitFlags::<AccessFs>::empty();
+        if perms & Perms::READ != Perms::empty() {
+            access |= AccessFs::from_read(ABI);
+        }
+        if perms & Perms::READ != Perms::empty() {
```

Comment:
This should be `Perms::WRITE`. Is it possible to add a test for this kind of `Perms` mapping?

Right now, all modules allowing `Perms::READ` permissions are also allowing write access, so we should check that the current policy is correct.

Maybe the `Perms` type could just be replaced with `BitFlags<AccessFs>` and a `TryFrom<&str>` implementation to manage the "rw" string mapping? This should simplify the code. We can still use a global `ACCESS_FS_READ = AccessFs::from_read(ABI);` (instead of `Perms::READ`) variable to simplify module configuration though.